L1Trigger/L1THGCal/src/veryfrontend/HGCalVFECompressionImpl.cc

```diff
@@ -1,5 +1,87 @@
 #include "L1Trigger/L1THGCal/interface/veryfrontend/HGCalVFECompressionImpl.h"
 
 HGCalVFECompressionImpl::
-HGCalVFECompressionImpl(const edm::ParameterSet& conf)
-{}
+HGCalVFECompressionImpl(const edm::ParameterSet& conf):
+    exponentBits_(conf.getParameter<uint32_t>("exponentBits")),
+    mantissaBits_(conf.getParameter<uint32_t>("mantissaBits")),
+    rounding_(conf.getParameter<bool>("rounding"))
+{
+  saturable_ = false;
+  saturationValue_ = 0xffffffff;
+  if (((1 << exponentBits_) + mantissaBits_ - 1) < 32) {
+    saturable_ = true;
+    saturationValue_ = (1 << ((1 << exponentBits_) + mantissaBits_ - 1)) - 1;
+  }
+  // TODO: for non-saturating need to get maximum code point as well?
+  // i.e. 8-bit compressed code that would hit 0xffffffff
+
+  // build the compressed values lookup table
+  uint32_t compval = 0;
+  const uint32_t mantissaMask = (1 << mantissaBits_) - 1;
+  uint32_t incval = 1;
+  for (int codeInt = 0; codeInt < 0x100; codeInt++) {
+    const uint8_t code = static_cast<uint8_t>(codeInt);
+    compressedValueLUT_[code] = compval;
+    if ((code & mantissaMask) == 0) {
+      incval <<= 1;
+      if (((code & ~mantissaMask) >> mantissaBits_) <= 1) {
+        incval >>= 1;
+      }
+    }
+    compval += incval;
+  }
+}
+
+uint8_t
+HGCalVFECompressionImpl::
+compressSingle(const uint32_t value)
+{
+  // check for saturation
+  if (saturable_ && value > saturationValue_)
+      return 0xff;
```

Comment:
The return saturation value should be equal to the saturation value, not `0xff`

Reply:
Maybe a special 32-bit constant for saturated conditions? Like:
`static const uint32_t saturatedValue = 0xffffffff;`
in the class declaration.

Reply:
No, it should be the maximum value that can be encoded on the exponent+mantissa word.